### PR TITLE
Cleanup Note

### LINKED ISSUE #111: [TASK] Refactor `Note`
Have a NoteBuilder that builds JsNote:
```rust
struct NoteBuilder {
    pub version: Option<NoteVersion>,
    pub secrets: Option<Vec<u8>>,
   ...
}

impl NoteBuilder {
	pub fn version(&mut self, version: Version) {
		let c: String = JsValue::from(&version).as_string().unwrap();
		self.version = Some(c.parse().unwrap());
	}

	#[wasm_bindgen(js_name= setSecrets)]
	pub fn set_secrets(&mut self, secrets: JsString) {
		let secrets_string: String = secrets.into();
		let sec = hex::decode(secrets_string.replace("0x", "")).unwrap();
		self.note_builder.secrets = Some(sec);
	}

      pub fn build(self) -> Result<JsNote, OpStatusCode> {
          if self.version.is_none() {
              return Err(OpStatusCode::InvalidNoteVersion);
          }

          let version = self.version.unwrap();

          return JsNote::new(version, ...);
      }
}
```

We should have two versions of the note, the rust version and wasm_bindgen version to be used by Javascript:
Rust Version:
```rust
pub struct Note {
	pub prefix: NotePrefix,
	pub version: NoteVersion,
	pub chain_id: String,
	pub source_chain_id: String,

	/// zkp related items
	pub backend: Backend,
	pub hash_function: HashFunction,
	pub curve: Curve,
	pub exponentiation: i8,
	pub width: usize,
	/// mixer related items
	pub secret: Vec<u8>,

	pub token_symbol: String,
	pub amount: String,
	pub denomination: u8,
}

impl Note {
    // Getters and Setters
   fn serialize(&self) -> String;
   fn deserialize(note_str: String) -> Note;
}
```
Javascript version -- just a wrapper around Note:
```rust
#[wasm_bindgen]
struct JsNote {
    note: Note,
}

impl JsNote {
    // Getters
   fn serialize(&self) -> JsString;
   fn deserialize(note_str: JsString) -> JsNote;
}
```

These types should all be living in one folder with structure:
`note/mod.rs` -- should contain Note and JsNote.
`note/builder.rs` -- should contain NoteBuilder

### PR DESCRIPTION
**Pull request by AhmedKorim**
Closes #111 , related #104

### CODE REVIEW

**Comment by AhmedKorim** [comment]
As @filiplazovic  recommended

*   Removal of `Arkworks_Posidedon_**` files and replace them with functions to generate the secrets
*   Removal of `width` of the note as the `exponentiation` is only needed
*   Moving `wasm` code related to note to `mod.rs` , and similarly for the prooving code

File: packages/wasm-utils/src/note/secrets.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		(..) => {
+			unreachable!("unreachable exponentiation {}", exponentiation)
+		}
+	};
+	Ok(secrets)
+}
+pub fn get_leaf_with_private_raw(curve: Curve, _exponentiation: i8, raw: &[u8]) -> (Vec<u8>, Vec<u8>) {
+	let secrets = raw[..32].to_vec();
+	let nullifer = raw[32..64].to_vec();
+	// todo consume the exponentiation parameters if needed in the future
```
Comment (drewstone) [inline packages/wasm-utils/src/note/secrets.rs:32]:
```suggestion
	// TODO: consume the exponentiation parameters if needed in the future
```

File: packages/wasm-utils/src/note/secrets.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			[secret_bytes, nullifier_bytes].concat()
+		}
+		(..) => {
+			unreachable!("unreachable exponentiation {}", exponentiation)
+		}
+	};
+	Ok(secrets)
+}
+pub fn get_leaf_with_private_raw(curve: Curve, _exponentiation: i8, raw: &[u8]) -> (Vec<u8>, Vec<u8>) {
+	let secrets = raw[..32].to_vec();
```
Comment (drewstone) [inline packages/wasm-utils/src/note/secrets.rs:30]: We should also ensure that `raw.len() >= 64`

File: packages/wasm-utils/src/wasm/mod.rs [hunk 3]
```diff
@@ -46,91 +53,170 @@ const VERSION: &str = "type Version = 'v1'";
...
 #[wasm_bindgen]
-pub struct NoteBuilderInput {
-	note_builder: NoteBuilder,
+pub struct JsNoteBuilder {
+	#[wasm_bindgen(skip)]
+	pub prefix: Option<NotePrefix>,
+	#[wasm_bindgen(skip)]
+	pub version: Option<NoteVersion>,
+	#[wasm_bindgen(skip)]
+	pub chain_id: Option<String>,
```
Comment (drewstone) [inline packages/wasm-utils/src/wasm/mod.rs:70]: can we call this `target_chain_id`?

**Review by drewstone — CHANGES_REQUESTED** [review]

File: packages/wasm-utils/src/note/secrets.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+use arkworks_gadgets::prelude::ark_bls12_381::Fq as BlsFr;
+use arkworks_gadgets::prelude::ark_bn254::Fq as Bn254Fr;
+use arkworks_gadgets::prelude::*;
+use arkworks_utils::utils::common::Curve as ArkworksCurve;
+
+use crate::types::{Curve, OpStatusCode};
+use crate::utils::get_hash_params_x5;
+
+pub fn generate_secrets(exponentiation: usize, curve: Curve, rng: &mut OsRng) -> Result<Vec<u8>, OpStatusCode> {
+	let secrets: Vec<u8> = match (curve, exponentiation) {
```
Comment (filiplazovic) [inline packages/wasm-utils/src/note/secrets.rs:12]: After the chat yesterday, we should bring back the `width` here.
The `width` in the note is regarding the leaf width, so it should be `5` here

File: packages/wasm-utils/src/note/secrets.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		}
+	};
+	Ok(secrets)
+}
+pub fn get_leaf_with_private_raw(curve: Curve, _exponentiation: i8, raw: &[u8]) -> (Vec<u8>, Vec<u8>) {
+	let secrets = raw[..32].to_vec();
+	let nullifer = raw[32..64].to_vec();
+	// TODO: consume the exponentiation parameters if needed in the future
+	return match curve {
+		Curve::Bls381 => {
```
Comment (filiplazovic) [inline packages/wasm-utils/src/note/secrets.rs:34]: We should also check for exponentiation and width here

File: packages/wasm-utils/src/note/secrets.rs (hunk 4 above)
Comment (AhmedKorim) [reply packages/wasm-utils/src/note/secrets.rs:12]: The leaf is always of width 5, for artworks, right? if not please say more about what makes width useful here, as generating secrets will just create the leaf

Comment (filiplazovic) [reply packages/wasm-utils/src/note/secrets.rs:12]: The leaf could be generated using widths other than 5 (for example with the width of 3)

File: packages/wasm-utils/src/note/secrets.rs (hunk 2 above)
Comment (AhmedKorim) [reply packages/wasm-utils/src/note/secrets.rs:47]: Added this

File: packages/wasm-utils/src/note/secrets.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+use crate::utils::get_hash_params_x5;
+
+pub fn generate_secrets(
+	exponentiation: i8,
+	width: usize,
+	curve: Curve,
+	rng: &mut OsRng,
+) -> Result<Vec<u8>, OpStatusCode> {
+	let secrets: Vec<u8> = match (curve, exponentiation, width) {
+		(Curve::Bls381, 5, ..) => {
```
Comment (filiplazovic) [inline packages/wasm-utils/src/note/secrets.rs:18]: We should also require that the leaf width is 5
```suggestion
		(Curve::Bls381, 5, 5) => {
```

File: packages/wasm-utils/src/types.rs [hunk 7]
```diff
@@ -157,4 +193,6 @@ pub enum OpStatusCode {
 	InvalidDenomination = 16,
 	/// Failed to generate secrets
 	SecretGenFailed = 17,
+	/// Invalid Note prefix
+	InvalidNotePrfix = 18,
```
Comment (filiplazovic) [inline packages/wasm-utils/src/types.rs:197]:
```suggestion
	InvalidNotePrefix = 18,
```

File: packages/wasm-utils/src/wasm/mod.rs [hunk 8]
```diff
@@ -46,91 +53,171 @@ const VERSION: &str = "type Version = 'v1'";
...
+			target_chain_id: self.target_chain_id.unwrap(),
+			source_chain_id: self.source_chain_id.unwrap(),
+			backend: self.backend.unwrap(),
+			hash_function: self.hash_function.unwrap(),
+			curve: self.curve.unwrap(),
+			token_symbol: self.token_symbol.unwrap(),
+			amount: self.amount.unwrap(),
+			denomination: self.denomination.unwrap(),
+			exponentiation: self.exponentiation.unwrap(),
+			width: self.width.unwrap(),
```
Comment (filiplazovic) [inline packages/wasm-utils/src/wasm/mod.rs:217]: Ideally, we want to return proper error here, instead of just doing `unwrap`

**Review by filiplazovic — APPROVED** [review]

**Pull request merged**